### PR TITLE
Migrate io.zeebe.broker.engine.job tests

### LINKED ISSUE #2556: Migrate `io.zeebe.broker.engine.job.*` tests
Migrate all tests which are part of the package `io.zeebe.broker.engine.job` to the engine module and move them in a corresponding package, related to the processor.

blocks #2461

### PR DESCRIPTION
**Pull request by miguelpires**
closes #2556

### CODE REVIEW

File: engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java [hunk 1]
```diff
@@ -0,0 +1,258 @@
+/*
```
[the remark this answers is not in the document]
Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java]: I think in this test you do not need to activate the jobs before. Jobs can also be completed without activation

File: engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+        engineRule.jobs().withType(JOB_TYPE).activateAndWait();
+    final Long jobKey = batchRecord.getValue().getJobKeys().get(0);
+    engineRule.job().fail(jobKey);
+
+    engineRule
+        .job()
+        .withVariables(
+            MsgPackUtil.asMsgPack(batchRecord.getValue().getJobs().get(0).getVariables()))
+        .complete(jobKey);
+    final Record<JobRecordValue> rejection =
```
Comment (Zelldon) [inline engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java]: Please take a look at the `PublishMessageClient` or at other clients how I implement the expected rejection

File: engine/src/test/java/io/zeebe/engine/job/FailJobTest.java [hunk 3]
```diff
@@ -0,0 +1,262 @@
+/*
```
Comment (Zelldon) [inline engine/src/test/java/io/zeebe/engine/job/FailJobTest.java]: Please move the job tests to `io.zeebe.engine.processor.workflow.job`

Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/job/FailJobTest.java]: I think in this test you do not need to activate the jobs before. Jobs can also be failed without activation

File: engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java [hunk 4]
```diff
@@ -15,58 +15,49 @@
...
-  public void setup() {
-    client = apiRule.partitionClient();
-  }
+  @Rule public EngineRule engineRule = new EngineRule();
 
   @Test
   public void shouldNotTimeOutIfDeadlineNotExceeded() {
     // given
-    brokerRule.getClock().pinCurrentTime();
+    engineRule.getClock().pinCurrentTime();
```
Comment (Zelldon) [inline engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java]: Please let us not pin the time, let us just avoid this. I see no benefit in this

File: engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java [hunk 5]
```diff
@@ -15,58 +15,49 @@
...
+    engineRule.createJob(jobType, PROCESS_ID);
 
-    apiRule.activateJobs(apiRule.getDefaultPartitionId(), jobType, timeout.toMillis()).await();
-    client.receiveFirstJobEvent(ACTIVATED);
+    engineRule.jobs().withType(jobType).withTimeout(timeout.toMillis()).activate();
+    jobRecords(ACTIVATED).getFirst();
 
     // when
-    brokerRule.getClock().addTime(timeout.minus(Duration.ofSeconds(1)));
+    engineRule.getClock().addTime(timeout.minus(Duration.ofSeconds(1)));
```
Comment (Zelldon) [inline engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java]: There is a utility in the engine rule already for this. I had my reasons why I don't exposed the clock :see_no_evil:

File: engine/src/test/java/io/zeebe/engine/job/FailJobTest.java (hunk 3 above)
Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/job/FailJobTest.java]: Could you try to migrate the jobs tests to ClassRules, please?

File: engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java [hunk 6]
```diff
@@ -75,18 +66,18 @@ public void shouldNotTimeOutIfDeadlineNotExceeded() {
   @Test
   public void shouldNotTimeOutIfJobCompleted() {
```
Comment (Zelldon) [inline engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java]: These kind of tests are terrible. They will take at least the max timeout time of the RecordingExporter, which is currently 5 sec.

File: engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java [hunk 7]
```diff
@@ -95,18 +86,20 @@ public void shouldNotTimeOutIfJobCompleted() {
   @Test
   public void shouldNotTimeOutIfJobFailed() {
     // given
-    brokerRule.getClock().pinCurrentTime();
+    engineRule.getClock().pinCurrentTime();
     final Duration timeout = Duration.ofSeconds(60);
     final String jobType = "jobType";
 
-    createJob(jobType);
+    engineRule.createJob(jobType, PROCESS_ID);
```
Comment (Zelldon) [inline engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java]: This test is flaky

File: engine/src/test/java/io/zeebe/engine/job/JobUpdateRetriesTest.java [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+    engineRule.createJob(JOB_TYPE, PROCESS_ID);
+    final Record<JobBatchRecordValue> batchRecord =
+        engineRule.jobs().withType(JOB_TYPE).activateAndWait();
+    final JobRecordValue job = batchRecord.getValue().getJobs().get(0);
+    final long jobKey = batchRecord.getValue().getJobKeys().get(0);
+
+    engineRule.job().withRetries(0).fail(jobKey);
+
+    // when
+    final Record<JobRecordValue> updatedRecord =
```
Comment (Zelldon) [inline engine/src/test/java/io/zeebe/engine/job/JobUpdateRetriesTest.java]: You can update the retires without fail and activate before right?

File: engine/src/test/java/io/zeebe/engine/util/EngineRule.java [hunk 9]
```diff
@@ -139,6 +143,10 @@ public void increaseTime(Duration duration) {
     environmentRule.getClock().addTime(duration);
   }
 
+  public ControlledActorClock getClock() {
```
Comment (Zelldon) [inline engine/src/test/java/io/zeebe/engine/util/EngineRule.java]: Please remove this again

File: engine/src/test/java/io/zeebe/engine/util/JobActivationClient.java [hunk 10]
```diff
@@ -71,14 +71,16 @@ public JobActivationClient withMaxJobsToActivate(int count) {
     return this;
   }
 
-  public Record<JobBatchRecordValue> activate() {
-    final long position =
-        environmentRule.writeCommandOnPartition(
-            partitionId, JobBatchIntent.ACTIVATE, jobBatchRecord);
-
+  public Record<JobBatchRecordValue> activateAndWait() {
```
Comment (Zelldon) [inline engine/src/test/java/io/zeebe/engine/util/JobActivationClient.java]: See my comment above about the conjunctions in method names

File: engine/src/test/java/io/zeebe/engine/util/JobClient.java [hunk 11]
```diff
@@ -71,31 +81,53 @@ public JobClient withType(String jobType) {
...
         .withSourceRecordPosition(position)
         .getFirst();
   }
 
-  public Record<JobRecordValue> fail() {
+  public long complete(long jobKey) {
+    return environmentRule.writeCommand(jobKey, JobIntent.COMPLETE, jobRecord);
+  }
+
+  public Record<JobRecordValue> failAndWait() {
```
Comment (Zelldon) [inline engine/src/test/java/io/zeebe/engine/util/JobClient.java]: This is not used

File: protocol-impl/src/main/java/io/zeebe/protocol/impl/record/value/job/JobBatchRecord.java [hunk 12]
```diff
@@ -127,4 +130,40 @@ public JobBatchRecord setTruncated(boolean truncated) {
...
+  @Override
+  public List<Long> getJobKeys() {
+    return StreamSupport.stream(jobKeysProp.spliterator(), false)
+        .map(LongValue::getValue)
+        .collect(Collectors.toList());
+  }
+
+  @Override
+  public List<JobRecordValue> getJobs() {
+    return StreamSupport.stream(jobsProp.spliterator(), false).collect(Collectors.toList());
```
Comment (Zelldon) [inline protocol-impl/src/main/java/io/zeebe/protocol/impl/record/value/job/JobBatchRecord.java]: I'm not sure how this works but I assume this will not create a copy right? If this is the case then we will have problems here, because if we reference these jobs and read a new JobBatchRecord the referenced job will change since the buffer is reused by the buffers

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Please see my comments below

File: engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java [hunk 13]
```diff
@@ -0,0 +1,258 @@
...
+  private static final String PROCESS_ID = "process";
+
+  @Rule public EngineRule engineRule = new EngineRule();
+
+  @Test
+  public void shouldCompleteJob() {
+    // given
+    engineRule.createJob(JOB_TYPE, PROCESS_ID);
+    final Record<JobBatchRecordValue> batchRecord =
+        engineRule.jobs().withType(JOB_TYPE).activateAndWait();
```
[the remark this answers is not in the document]
Comment (MiguelPires) [reply engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java]: In some tests I don't want to wait for the ACTIVATED or COMPLETED record because the command is supposed to be rejected. That's the only reason why I separated the writing of the command and the waiting for the record. In that case, I'll change this into an `.await()` method that waits for the follow-up event so we avoid having multiple things happening in the method with a conjunction. Do you agree?

[the remark this answers is not in the document]
Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java]: Ok got it. This is the reason why I added on some clients the method `expectRejection()`, which then waits until the rejeciton published instead of the follow up event. https://github.com/zeebe-io/zeebe/blob/develop/engine/src/test/java/io/zeebe/engine/util/PublishMessageClient.java#L99 or https://github.com/zeebe-io/zeebe/blob/74129fa41b14f645a01d59bfc357c963a98890b7/engine/src/test/java/io/zeebe/engine/util/DeploymentClient.java#L120

File: engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (MiguelPires) [reply engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java]: I think the previous tests did that to use the ACTIVATED record to assert job properties https://github.com/zeebe-io/zeebe/blob/6ff7796d2111e4c6a3e1f92674a266dd3b099464/broker-core/src/test/java/io/zeebe/broker/engine/job/CompleteJobTest.java#L62
But I can use the created record for the same purpose then and remove the activation stuff

File: engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (MiguelPires) [reply engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java]: Ah I hadn't noticed that, that's a better solution. Thanks!

File: engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java]: Lets keep the activation stuff before the complete and fail. Please create a new issue where we discuss if a job should be activated before complete/fail or not.

[the remark this answers is not in the document]
Comment (MiguelPires) [reply engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java]: Sure, here it is https://github.com/zeebe-io/zeebe/issues/2600

[the remark this answers is not in the document]
Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/job/CompleteJobTest.java]: Thanks :+1:

File: engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java (hunk 6 above)
Comment (MiguelPires) [reply engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java]: I know but it was already there and I didn't want to remove it unilaterally. Should I?

Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java]: Sorry, didn't wanted to blame you was more a comment :sweat_smile: Yes maybe better. I assume in the JobStateTest it is tested that completed or failed jobs are not available anymore.

Comment (MiguelPires) [reply engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java]: No worries. Sure thing, I'll remove it

File: engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java (hunk 4 above)
Comment (MiguelPires) [reply engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java]: Currently I need some way to test that the deadline is correct and I think for that I need to pin time. Something like this:
```
    final ControlledActorClock clock = engineRule.getClock();
    clock.pinCurrentTime();

    final Duration timeout = Duration.ofMinutes(4);
    final Instant deadline = clock.getCurrentTime().plus(timeout);
```

And then I can check the record against this deadline. Do you know if I can do this in another way without pinning and exposing the clock or do I have to remove this assertion?

Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java]: First of all this test tries to test again something which should not happen.

Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java]: And each record has a timestamp which you could use

File: protocol-impl/src/main/java/io/zeebe/protocol/impl/record/value/job/JobBatchRecord.java (hunk 12 above)
Comment (MiguelPires) [reply protocol-impl/src/main/java/io/zeebe/protocol/impl/record/value/job/JobBatchRecord.java]: This creates a new list with the JobRecords in jobProp. Just to be sure, what you're saying is that if we reuse the batch record the referenced jobs will change? Because I'd assume that the JobRecords to which the returned list points wouldn't change

File: engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java (hunk 4 above)
Comment (MiguelPires) [reply engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java]: Ok my bad, this is the JobTimeOutTest and I was looking at the ActivateJobsTest. In the activate test, I have the setup I described above and then I assert that the record's deadline is correct. How can I test the activated record without using the clock?

File: engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java (hunk 6 above)
Comment (MiguelPires) [reply engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java]: This class has several tests that use `assertNoMoreJobsReceived`, which slows them considerably. Should I removed them as well? I'd be removing half of the remaining tests

Comment (Zelldon) [reply engine/src/test/java/io/zeebe/engine/job/JobTimeOutTest.java]: Remove them, please. Except you see any value in them. I will take a look on the second review as well

**Comment by MiguelPires** [comment]
Thanks for the feedback @Zelldon. I improved the job client with expectation suppliers like you did on message client, I removed the tests that waited 5 seconds (the whole package now takes 2 seconds to run) and migrated the tests to class rules. However, I didn't remove the clock stuff from one ActivateJobsTest because I wasn't sure how to do that and still assert what I needed. Let me know what you think

File: protocol-impl/src/main/java/io/zeebe/protocol/impl/record/value/job/JobBatchRecord.java [hunk 14]
```diff
@@ -159,7 +159,20 @@ public int getMaxJobsToActivate() {
...
+
+    return StreamSupport.stream(jobsProp.spliterator(), false)
+        .map(
+            r ->
+                new JobRecord()
+                    .setRetries(r.getRetries())
+                    .setDeadline(r.getDeadlineLong())
+                    .setErrorMessage(r.getErrorMessageBuffer())
+                    .setWorker(r.getWorkerBuffer())
+                    .setVariables(r.getVariablesBuffer())
```
Comment (Zelldon) [inline protocol-impl/src/main/java/io/zeebe/protocol/impl/record/value/job/JobBatchRecord.java]: We have still here the problem that the buffers are wrapped, which means they use internally the same reference.

Comment (Zelldon) [reply protocol-impl/src/main/java/io/zeebe/protocol/impl/record/value/job/JobBatchRecord.java]: We have to create a new byte array and write the values into this array. After that we can wrap this array. Here is the code how I did it on the DeploymentRecord for the worfklows

```
  @Override
  public List<DeployedWorkflow> getDeployedWorkflows() {
    final List<DeployedWorkflow> workflows = new ArrayList<>();

    final Iterator<Workflow> iterator = workflowsProp.iterator();
    while (iterator.hasNext()) {
      final Workflow workflow = iterator.next();

      final byte[] bytes = new byte[workflow.getLength()];
      final UnsafeBuffer copyBuffer = new UnsafeBuffer(bytes);
      workflow.write(copyBuffer, 0);

      final Workflow copiedWorkflow = new Workflow();
      copiedWorkflow.wrap(copyBuffer);
      workflows.add(copiedWorkflow);
    }

    return workflows;
  }
```

File: engine/src/test/java/io/zeebe/engine/util/JobClient.java [hunk 15]
```diff
@@ -17,25 +17,65 @@
...
+              .withSourceRecordPosition(position)
+              .getFirst();
+
+  private static final Function<Long, Record<JobRecordValue>> UPDATE_RETRIES_SUCCESS_SUPPLIER =
+      (position) ->
+          RecordingExporter.jobRecords(JobIntent.RETRIES_UPDATED)
+              .withSourceRecordPosition(position)
+              .getFirst();
+
+  private static final Function<Long, Record<JobRecordValue>> UPDATE_RETRIES_REJECTION_SUPPLIER =
```
Comment (Zelldon) [inline engine/src/test/java/io/zeebe/engine/util/JobClient.java]: I think they can be summarized as:

```
private static final Function<Long, Record<JobRecordValue>> SUCCESS_SUPPLIER =
      (position) ->
          RecordingExporter.jobRecords()
              .onlyEvents()
              .withSourceRecordPosition(position)
              .getFirst();

  private static final Function<Long, Record<JobRecordValue>> REJECTION_SUPPLIER =
      (position) ->
          RecordingExporter.jobRecords()
              .onlyCommandRejections()
              .withSourceRecordPosition(position)
              .getFirst();

```

Then you do not need to distinguish between them

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Nice. Thanks @MiguelPires, just two minor notes.

**Comment by MiguelPires** [comment]
Thanks for the feedback @Zelldon. I fixed the issues you mentioned

File: engine/src/test/java/io/zeebe/engine/util/JobClient.java [hunk 16]
```diff
@@ -112,17 +82,17 @@ public JobClient withErrorMessage(String message) {
   }
 
   public JobClient expectRejectionOnComplete() {
-    completeSupplier = COMPLETE_REJECTION_SUPPLIER;
+    completeSupplier = REJECTION_SUPPLIER;
```
Comment (Zelldon) [inline engine/src/test/java/io/zeebe/engine/util/JobClient.java]: They can also be summarized then with `expectRejection()` and you only have on field `expectation`, which is then overwritten and used in the update, fail or complete

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]

File: exporter-api/src/main/java/io/zeebe/exporter/api/record/value/JobBatchRecordValue.java [hunk 17]
```diff
@@ -36,7 +36,7 @@
    *     exceeded, it can happen that the job is handed to another worker and the work is performed
    *     twice.
    */
-  Duration getTimeout();
+  Duration getTimeoutDuration();
```
Comment (Zelldon) [inline exporter-api/src/main/java/io/zeebe/exporter/api/record/value/JobBatchRecordValue.java]: Please revert this change - the plan was not to change the API more the implementations

Comment (Zelldon) [reply exporter-api/src/main/java/io/zeebe/exporter/api/record/value/JobBatchRecordValue.java]: So please rename to `JobBatchRecord#getTimeout` to `#getTimeoutLong` instead, otherwise we have here a breaking API change

**Review by ChrisKujawa — CHANGES_REQUESTED** [review]
Sorry overseen this before

**Comment by Zelldon** [comment]
If you want I can also finish and rebase the branch

**Comment by MiguelPires** [comment]
Ok, fixed those issues. It's ok, I'll squash and rebase. Thanks though

**Review by ChrisKujawa — APPROVED** [review]
Thanks @MiguelPires
